cmd/main.go

```diff
@@ -4,137 +4,151 @@ import (
-	flag.StringVar(&logLevel, "logLevel", "", "Log level (debug, info, warn, error, dpanic, panic, fatal)")
+	flag.StringVar(&cliListeners, "listeners", "", "Comma-separated list of listeners to override config")
+	flag.StringVar(&logLevel, "logLevel", "", "Log level")
 	flag.StringVar(&certFile, "certFile", "", "Path to the certificate file")
 	flag.StringVar(&keyFile, "keyFile", "", "Path to the key file")
-	flag.BoolVar(&useTLS, "useTLS", true, "Whether to use TLS")
-
+	flag.BoolVar(&useTLS, "useTLS", false, "Whether to use TLS")
 	flag.Parse()
 
-	// Load configuration
 	cfg, err := internal.LoadConfig(configPath)
 	if err != nil {
-		panic("Couldn't load configuration")
-	}
-	if logLevel != "" {
-		cfg.LogLevel = logLevel
+		log.Fatal("Couldn't load configuration: ", err)
 	}
+
+	cfg.UseTLS = useTLS
+
 	if certFile != "" {
 		cfg.CertFile = certFile
 	}
+
+	if logLevel != "" {
+		cfg.LogLevel = logLevel
+	}
+
 	if keyFile != "" {
 		cfg.KeyFile = keyFile
 	}
-	// TLS is enabled by default, but can be disabled with a flag for testing
-	if !useTLS {
-		cfg.UseTLS = useTLS
-	}
-	fmt.Println("Loaded log level from config:", cfg.LogLevel)
-
-	// Validate upstreams and collect priorities
-	prioritySet := make(map[int]bool)
-	for name, upstream := range cfg.Upstreams {
-		if upstream.Type != "azure" && upstream.Type != "openai" {
-			fmt.Printf("Invalid type for upstream %s: %s\n", name, upstream.Type)
-			return
-		}
-		if upstream.Type == "azure" && upstream.URL == "" {
-			fmt.Printf("Missing URL for Azure upstream %s\n", name)
-			return
-		}
-		if upstream.Type == "openai" && upstream.URL != "" {
-			fmt.Printf("URL should not be provided for OpenAI upstream %s\n", name)
-			return
-		}
-		if upstream.Priority <= 0 {
-			fmt.Printf("Invalid priority for upstream %s: %d\n", name, upstream.Priority)
-			return
-		}
-		if _, exists := prioritySet[upstream.Priority]; exists {
-			fmt.Printf("Duplicate priority for upstream %s: %d\n", name, upstream.Priority)
-			return
-		}
-		prioritySet[upstream.Priority] = true
-	}
 
-	// If listeners are provided via CLI, override the ones from the config
 	if cliListeners != "" {
-		cfg.Listeners = []internal.Listener{}
-		for _, cliListener := range strings.Split(cliListeners, ",") {
-			parts := strings.Split(cliListener, ":")
-			if len(parts) != 2 {
-				fmt.Println("Invalid listener format, skipping:", cliListener)
-				continue
-			}
-			cfg.Listeners = append(cfg.Listeners, internal.Listener{Interface: parts[0], Port: parts[1]})
-		}
+		overrideListeners(cfg, cliListeners)
 	}
 
-	// Initialize logger
-	logger, err := internal.InitializeLogger(cfg.LogLevel)
+	// Initialize the logger in logger.go
+	logger, err := internal.InitializeLogger(logLevel)
 	if err != nil {
-		panic("Couldn't initialize logger")
+		log.Fatal("Couldn't initialize logger: ", err)
 	}
-	defer logger.Sync()
+	ctx := createContext(cfg, logger)
```

Comment:
You should create the context here, not in the called function. Contexts are passed in, not returned. This is an anti pattern.
```suggestion
	ctx := createContext(context.Background(), cfg, logger)
```

Then modify the called function to take in a context instead of create one.